src/core/connection.c

```diff
@@ -5512,24 +5512,40 @@ QuicConnRecvDatagrams(
 }
 
 _IRQL_requires_max_(PASSIVE_LEVEL)
-void
+BOOLEAN
 QuicConnFlushRecv(
     _In_ QUIC_CONNECTION* Connection
     )
 {
+    BOOLEAN FlushedAll;
     uint32_t ReceiveQueueCount;
     CXPLAT_RECV_DATA* ReceiveQueue;
 
     CxPlatDispatchLockAcquire(&Connection->ReceiveQueueLock);
-    ReceiveQueueCount = Connection->ReceiveQueueCount;
-    Connection->ReceiveQueueCount = 0;
     ReceiveQueue = Connection->ReceiveQueue;
-    Connection->ReceiveQueue = NULL;
-    Connection->ReceiveQueueTail = &Connection->ReceiveQueue;
+    if (Connection->ReceiveQueueCount > QUIC_MAX_RECEIVE_FLUSH_COUNT) {
+        FlushedAll = FALSE;
+        Connection->ReceiveQueueCount -= QUIC_MAX_RECEIVE_FLUSH_COUNT;
+        CXPLAT_RECV_DATA* Tail = Connection->ReceiveQueue;
+        ReceiveQueueCount = 0;
+        while (ReceiveQueueCount++ < QUIC_MAX_RECEIVE_FLUSH_COUNT) {
+            Tail = Connection->ReceiveQueue;
+            Connection->ReceiveQueue = Connection->ReceiveQueue->Next;
+        }
```

Comment:
I think this while loop can be optimized a bit (mainly for better readability), isn't it?

```
        while (++ReceiveQueueCount < QUIC_MAX_RECEIVE_FLUSH_COUNT) {
            Tail = Tail->Next;
        }
        Connection->ReceiveQueue = Tail->Next;
        Tail->Next = NULL;
```

Reply:
Yeah, I think that works.